splice_machine/src/main/java/com/splicemachine/derby/stream/function/merge/AbstractMergeJoinFlatMapFunction.java

```diff
@@ -112,56 +113,103 @@ private boolean isKeyColumn(int[] columnOrdering, int col) {
-                scanStartOverride = startPosition;
-            else {
-
-                ExecRow newStartKey = new ValueRow(columnOrdering.length);
-                int[] rightHashKeys = joinOperation.getRightHashKeys();
-
-                for (int i = 0; i < nCols; i++)
-                    newStartKey.setColumn(i + 1, startPosition.getColumn(i + 1));
-
-                int len = nCols;
-                for (int i = len; i < columnOrdering.length; i++) {
-                    int keyColumnBaseTablePosition = columnOrdering[i];
-                    int j = 0;
-                    for (; j < rightHashKeys.length; j++) {
-                        // both columnOrdering and colToBaseTableMap are 0-based
-                        if (colToBaseTableMap[j] == keyColumnBaseTablePosition) {
+        else {
+
+            ExecRow newStartKey = new ValueRow(columnOrdering.length);
+            int[] rightHashKeys = joinOperation.getRightHashKeys();
+            int[] rightHashKeySortOrders = ((MergeJoinOperation) joinOperation).getRightHashKeySortOrders();
+
+            int len = 0;
+            for (int i = 0; i < columnOrdering.length; i++) {
+                int keyColumnBaseTablePosition = columnOrdering[i];
+                int j = 0;
+                for (; j < rightHashKeys.length; j++) {
+                    // both columnOrdering and colToBaseTableMap are 0-based
+                    if (colToBaseTableMap[j] == keyColumnBaseTablePosition) {
+                        if (i >= nCols) {
                             newStartKey.setColumn(i + 1, firstHashRow.getColumn(j + 1));
-                            len++;
-                            break;
                         }
-                    }
-                    //no match found for the given key column position
-                    if (j >= rightHashKeys.length)
+                        else {
+
+                            boolean rightKeyIsNull = startPosition == null ||
+                                    isNullDataValue(startPosition.getColumn(i + 1));
+                            boolean leftKeyIsNull =
+                                    isNullDataValue(firstHashRow.getColumn(j + 1));
+
+                            // Replace the right key value to seek to with left row data value if:
+                            // - The right key is null, in which case left is always >= right , or
+                            // - If neither left key nor right key is null, compare
+                            //   left and right key values, and do the replacement if:
+                            //   - If the key is ascending and the current right key start
+                            //     position is less than the left data value, or
+                            //   - If the key is descending and the current right key start
+                            //     position is greater than the left data value.
+                            boolean replaceWithLeftRowVal = false;
```

Comment:
I'm thinking if replaceWithLeftRowVal should be a decision based on every column, or we should do it once for the leading index/pk columns, and let the subsequent columns honor the decision. 
For example, given firstHashRow (1,3), startPosition (2,1), and all columns in asc order, for the first column, we will pick 2 from startPosition, for the subsequent columns, we should continue to pick 1 from startPosition instead of 3 from firstHashRow. I think the existing logic also has an issue.
Here is an example:
create table t1 (a1 int, b1 int, c1 int, primary key (a1, b1));
create table t2 (a2 int, b2 int, c2 int, primary key (a2, b2));
insert into t1 values (1,3,3), (2,1,1);
insert into t2 values (1,3,3), (2,1,1);

select * from t1, t2 --splice-properties joinStrategy=merge
where a1=a2 and b1=b2 and a2>=4/2;

Reply:
In this case, since a2>=4/2 contains an expression, TC does not kick in, so the left right may not start from a1=2 but a row with a smaller value. I guess we cannot rely on TC to always be kicking in for the logic initRightScan().

Reply:
Good catch @yxia92.  I updated the code to keep comparing key columns as long as they are equal, and once a partial key is found to be greater than the other (left vs. right), we pick that one as the start key and stick with that decision.